### PR TITLE
Add IOI legacy and ICPC contest format; #79

### PR DESCRIPTION
**Pull request by Carson-Tang**

### CODE REVIEW

File: docs/site/contest_formats.md [hunk 1]
```diff
@@ -1,22 +1,31 @@
 # Contest Formats
 
-The DMOJ ships with 4 contest formats by default: Default, Codechef IOI Ranklist (henceforth shortened to simply IOI), ECOO, and AtCoder.
+The DMOJ ships with 6 contest formats by default: Default, Codechef IOI Ranklist (henceforth shortened to simply IOI), Legacy IOI (pre-2016), ECOO, AtCoder and ICPC.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:3]: Codechef IOI Ranklist is legacy IOI...

File: docs/site/contest_formats.md [hunk 2]
```diff
@@ -1,22 +1,31 @@
...
+The score is the sum of the highest-scoring submission on each problem, and ties are broken by adding the time of the last submission to each problem with a non-zero maximum score.
 
 Note that *any* submission will increase time penalty, not just score-changing submissions.
 
 There are no additional options that can be configured for this contest format.
 
 ## IOI Contest Format
 
-The `IOI` contest format emulates the scoring used by Codechef's IOI Ranklist.
+The `IOI` contest format emulates the scoring used by Codechef's IOI Ranklist from 2016 onwards.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:17]: Just IOI period, not Codechef IOI.

File: docs/site/contest_formats.md [hunk 3]
```diff
@@ -1,22 +1,31 @@
...
 
-The score is equal to the sum of the maximum scores on each problem, and by default, ties are not broken.
+The score is equal to the sum of the final score on each problem, where the final score for a problem is the maximum score for each subtask across all submissions, and by default, ties are not broken.
+For example, consider a contestant made two submissions on a task with two subtasks. The first gets 30 points on the first subtask and 10 points on the second subtask. The second gets 0 points on the first subtask and 40 points on the second subtask. The final score for this problem will be 70.
+
+The `cumtime` option can be set to `true` within the JSON configuration. This will break ties by summing the submissions times of the most recent submissions.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:22]: Are you sure it's "most recent", not "the first submissions that passes each subtask"

File: docs/site/contest_formats.md [hunk 4]
```diff
@@ -37,7 +46,19 @@ For example, say a submission with a score of `50/100` is submitted 23 minutes b
 
 ## AtCoder Contest Format
 
-The `AtCoder` contest format is based on the contest format used by AtCoder. As such, your score equals the sum of your highest-scoring submission on each problem, and ties are broken based on the time of your last score-changing submission plus your penalty.
+The `AtCoder` contest format is based on the contest format used by AtCoder.
+
+The score is equal to the sum of the highest-scoring submission on each problem, and ties are broken based on the time of the last score-changing submission plus the penalty.
 
-The penalty is specified by the `penalty` field, and defaults to 5 minutes.
+The penalty is specified by the `penalty` field, and defaults to `5` minutes.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:53]: I don't think it was necessary to put in a code tag

File: docs/site/contest_formats.md [hunk 5]
```diff
@@ -37,7 +46,19 @@ For example, say a submission with a score of `50/100` is submitted 23 minutes b
...
 
-The penalty is specified by the `penalty` field, and defaults to 5 minutes.
+The penalty is specified by the `penalty` field, and defaults to `5` minutes.
 The penalty is equal to the total number of incorrect submissions prior to the highest scoring submission on each solved problem, multiplied by the specified value, in minutes, and is added to the cumulative time.
+
+## ICPC Contest Format
+
+The `ICPC` contest format is based on the contest format used by the ICPC.
+
+The score is equal to the number of problems solved, and ties are broken based on the sum of the elapsed time that a correct submission was submitted to each problem plus the penalty.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:60]: There are 3 tiebreakers, third one is the time of last score-alterating submission iirc

File: docs/site/contest_formats.md [hunk 6]
```diff
@@ -37,7 +46,19 @@ For example, say a submission with a score of `50/100` is submitted 23 minutes b
...
+
+The penalty is specified by the `penalty` field, and defaults to `20` minutes.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:62]: No need to put the 20 in code tag

File: docs/site/contest_formats.md [hunk 7]
```diff
@@ -1,22 +1,31 @@
...
 There are no additional options that can be configured for this contest format.
 
 ## IOI Contest Format
 
-The `IOI` contest format emulates the scoring used by Codechef's IOI Ranklist.
+The `IOI` contest format emulates the scoring used by IOI.
 
-The score is equal to the sum of the maximum scores on each problem, and by default, ties are not broken.
+The score is equal to the sum of the final score on each problem, where the final score for a problem is the maximum score for each subtask across all submissions, and by default, ties are not broken.
+For example, consider a contestant made two submissions on a task with two subtasks. The first gets 30 points on the first subtask and 10 points on the second subtask. The second gets 0 points on the first subtask and 40 points on the second subtask. The final score for this problem will be 70.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:20]: > consider a contestant made two submissions on a task with two subtasks

grammar is wrong

File: docs/site/contest_formats.md [hunk 8]
```diff
@@ -1,22 +1,31 @@
...
+
+The `cumtime` option can be set to `true` within the JSON configuration. This will break ties by summing the submissions times of the first submissions that passes each subtask".
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:22]: Wait is this actually what happens? (I didn't read the code carefully last time, just took a guess)

File: docs/site/contest_formats.md [hunk 9]
```diff
@@ -28,16 +37,28 @@ The score is equal to the sum of scores of the last submissions to each problem.
...
 
 The `first_ac_bonus` field, as suggested by the name, will add the specified number of points to a problem's score if it is solved on the first attempt, excluding compile errors and internal errors.
-This field defaults to `10`.
+This field defaults to 10.
 
 The `time_bonus` field awards a bonus for solving problems faster.
-The field is in minutes, and for each such interval of time before the contest ends, any submission with a non-zero score will have a bonus added. This field defaults to `5`.  
+The field is in minutes, and for each such interval of time before the contest ends, any submission with a non-zero score will have a bonus added. This field defaults to 5.  
 Note that specifying `0` will disable this bonus.  
 For example, say a submission with a score of `50/100` is submitted 23 minutes before the contest ends. $\left\lfloor\frac{23}{5}\right\rfloor = 4$, so 4 bonus points are awarded, giving a total score of $50 + 4 = 54$ for that problem.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:45]: You can probably remove the code tag around 50/100 if you're going to clean this up :)

File: docs/site/contest_formats.md [hunk 10]
```diff
@@ -28,16 +37,28 @@ The score is equal to the sum of scores of the last submissions to each problem.
...
+
+## ICPC Contest Format
+
+The `ICPC` contest format is based on the contest format used by the ICPC.
+
+The score is equal to the number of problems solved, and ties are broken firstly based on the sum of the elapsed time that a correct submission was submitted to each problem plus the penalty and secondly based on the time of the last score-changing submission. 
+
+The penalty is specified by the `penalty` field, and defaults to 20 minutes.
+The penalty is equal to the number of incorrect submissions prior to the first correct submission, multiplied by the specified value, in minutes, and is added to the cumulative time. 
+Note that there is no time penalty for a problem that is not solved.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:64]: I believe you get penalty as long as you have a non-zero score (solved implies AC)

File: docs/site/contest_formats.md (hunk 8 above)
Comment (Carson-Tang) [reply docs/site/contest_formats.md:22]: That is what should be implemented.

File: docs/site/contest_formats.md (hunk 10 above)
Comment (Carson-Tang) [reply docs/site/contest_formats.md:64]: There are no partials.

Comment (kiritofeng) [reply docs/site/contest_formats.md:64]: Incorrect: this format can be used on contests that have partials enabled.

File: docs/site/contest_formats.md [hunk 11]
```diff
@@ -1,22 +1,31 @@
 # Contest Formats
 
-The DMOJ ships with 4 contest formats by default: Default, Codechef IOI Ranklist (henceforth shortened to simply IOI), ECOO, and AtCoder.
+The DMOJ ships with 6 contest formats by default: Default, IOI, Codechef IOI Ranklist (henceforth shortened to simply Legacy IOI), ECOO, AtCoder and ICPC.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:3]: Oxford comma pls

File: docs/site/contest_formats.md [hunk 12]
```diff
@@ -28,16 +37,28 @@ The score is equal to the sum of scores of the last submissions to each problem.
 By default, ties are not broken, however, setting `cumtime` to `true` will sum the times of the last submission to each problem, and use that to tiebreak.
 
 The `first_ac_bonus` field, as suggested by the name, will add the specified number of points to a problem's score if it is solved on the first attempt, excluding compile errors and internal errors.
-This field defaults to `10`.
+This field defaults to 10.
 
 The `time_bonus` field awards a bonus for solving problems faster.
-The field is in minutes, and for each such interval of time before the contest ends, any submission with a non-zero score will have a bonus added. This field defaults to `5`.  
+The field is in minutes, and for each such interval of time before the contest ends, any submission with a non-zero score will have a bonus added. This field defaults to 5.  
 Note that specifying `0` will disable this bonus.
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:44]: Should remove code tag around 0 to be consistent I guess...

File: docs/site/contest_formats.md [hunk 13]
```diff
@@ -28,16 +37,28 @@ The score is equal to the sum of scores of the last submissions to each problem.
...
+
+## ICPC Contest Format
+
+The `ICPC` contest format is based on the contest format used by the ICPC.
+
+The score is equal to the number of problems solved, and ties are broken firstly based on the sum of the elapsed time that a correct submission was submitted to each problem plus the penalty and secondly based on the time of the last score-changing submission. 
+
+The penalty is specified by the `penalty` field, and defaults to 20 minutes.
+The penalty is equal to the number of incorrect submissions prior to the first correct submission, multiplied by the specified value, in minutes, and is added to the cumulative time. 
+Note that there is time penalty for any non-zero score problem (this format will not automatically disable partial points).
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:64]: > there is time penalty for any non-zero score problem
try
> the time penalty is applied to all problems with a non-zero score

File: docs/site/contest_formats.md [hunk 14]
```diff
@@ -1,22 +1,31 @@
...
 ## IOI Contest Format
 
-The `IOI` contest format emulates the scoring used by Codechef's IOI Ranklist.
+The `IOI` contest format emulates the scoring used by IOI.
 
-The score is equal to the sum of the maximum scores on each problem, and by default, ties are not broken.
+The score is equal to the sum of the final score on each problem, where the final score for a problem is the maximum score for each subtask across all submissions, and by default, ties are not broken.
+For example, consider a contestant that makes two submissions on a task with two subtasks. The first gets 30 points on the first subtask and 10 points on the second subtask. The second gets 0 points on the first subtask and 40 points on the second subtask. The final score for this problem will be 70.
+
+The `cumtime` option can be set to `true` within the JSON configuration. This will break ties by summing the submissions times of the first submissions that passes each subtask".
```
Comment (kiritofeng) [inline docs/site/contest_formats.md:22]: Remove the quotation mark

**Pull request merged**